shell/common/persistent_cache.cc

```diff
@@ -79,6 +79,29 @@ void PersistentCache::SetCacheDirectoryPath(std::string path) {
 }
 
 namespace {
+
+constexpr char kEngineComponent[] = "flutter_engine";
+
+static void FreeOldCacheDirectory(const fml::UniqueFD& cache_base_dir) {
+  fml::UniqueFD engine_dir =
+      fml::OpenDirectoryReadOnly(cache_base_dir, kEngineComponent);
+  if (!engine_dir.is_valid()) {
+    return;
+  }
+  fml::VisitFiles(engine_dir, [](const fml::UniqueFD& directory,
+                                 const std::string& filename) {
+    if (filename != GetFlutterEngineVersion()) {
+      auto dir = fml::OpenDirectory(directory, filename.c_str(), false,
+                                    fml::FilePermission::kReadWrite);
+      if (dir.is_valid()) {
+        fml::RemoveFilesInDirectory(dir);
```

Comment:
It would be simpler to have the directory tree delete function take a parent `UniqueFD` and the name of the directory to delete (similar to `fml::UnlinkDirectory`, but recursive).

With that change the delete function can unlink the directory itself after all the contents have been removed.

Reply:
Good call.

Reply:
Done with `RemoveDirectoryRecursively` added. The old `RemoveFilesInDirectory` is still preserved for unit tests (e.g., https://github.com/flutter/engine/pull/18244/files#diff-b3bfc82230a960f6d96fe512bacd4d7cR235)